### ISSUE TITLE
How do I add a Data source and bind it to JNDI?

### ISSUE THREAD

**Issue by lsantos** [payara/docker-payaraserver-full#13]
Hi there,

I am not sure if this is the right place to ask this, but I would like to run these two commands right after payara start-domain runs:

RUN ${PAYARA_PATH}/bin/asadmin create-jdbc-connection-pool \
--datasourceclassname org.postgresql.ds.PGConnectionPoolDataSource \
--restype javax.sql.ConnectionPoolDataSource \
--property user=postgres:password=:DatabaseName=fin_dev:ServerName=$HOST_IP:port=5432 fing-pool

RUN ${PAYARA_PATH}/bin/asadmin create-jdbc-resource --connectionpoolid fing-pool jdbc/fing

How could I accomplish this? 

I was thinking about writing a bash script and creating my new image whose entry point would execute this bash file, but I am not sure if this is the right thing to do.

**Comment by mikecroft** [comment]
I suppose this is as relevant a place as any, though it's not quite so visible as [our Google Group forum](https://groups.google.com/forum/#!forum/payara-forum)!

You're fortunate to ask this now, because the 172 release has added a couple of features to make these things much more easy with Docker: the ability to run pre and post boot command files on startup.

Here's a preview of the docs which will be published properly tomorrow:
https://docs.payara.fish/v/next/documentation/extended-documentation/app-deployment/app-deployment.html


@OndrejM has modified the Dockerfile to take advantage of this feature so that any app that you `COPY` to `${PAYARA_PATH}/deployments` will be deployed using the new command file. Looking at the script that Ondrej used to generate the commands, these will be ***appended*** to the post boot file, so I believe you could create your own post boot file in the same place with your own commands, and these should be applied as well.

(this is based on a pretty massive assumption about how Docker works, I haven't tried this myself yet, it's all pretty new! - @OndrejM please correct me if I'm wrong!)

Basically, I think you should be able to add these two commands to a file (`${PAYARA_PATH}/post-boot-commands.asadmin`) something like this:

```
create-jdbc-connection-pool --datasourceclassname org.postgresql.ds.PGConnectionPoolDataSource --restype javax.sql.ConnectionPoolDataSource --property user=postgres:password=:DatabaseName=fin_dev:ServerName=$HOST_IP:port=5432 fing-pool

create-jdbc-resource --connectionpoolid fing-pool jdbc/fing
```

Then the new `ENTRYPOINT` will apply your post boot commands before then deploying anything in the deployments directory.

**Comment by OndrejM** [comment]
@mikecroft, you're mostly right, except the script to generate the file with asadmin commands always regenerates the contents. To add custom commands into the file, you could modify the ENTRYPOINT to copy the generated commands into a new file and use that one when Payara Server is started.

```
ENTRYPOINT=${PAYARA_PATH}/generate_deploy_commands.sh && \
  echo 'create-jdbc-connection-pool --datasourceclassname org.postgresql.ds.PGConnectionPoolDataSource
  --restype javax.sql.ConnectionPoolDataSource --property user=postgres:password=:DatabaseName=fin_dev:ServerName=$HOST_IP:port=5432 fing-pool' > mycommands.asadmin && \
  cat ${DEPLOY_COMMANDS} >> mycommands.asadmin && \
  ${PAYARA_PATH}/bin/asadmin start-domain -v --postbootcommandfile mycommands.asadmin ${PAYARA_DOMAIN}
```

This is not documented in our documentation, it's more a feature of Docker. We'll improve the docker images later to support this usecase better.

**Comment by OndrejM** [comment]
You can also modify the entrypoint when you run the docker image, with the `--entrypoint` parameter to the `docker run`. But it is more complicated - you can only provide an executable as the entrypoint and pass parameters to it at the end of the command line. So ideally you would set '/bin/bash' as the entry point and use the `-c` option to pass arguments to it like this:

```
docker run -i -t -p 4848:4848 -p 8080:8080 --entrypoint '/bin/bash'  payara/server-full \ 
  -c '${PAYARA_PATH}/generate_deploy_commands.sh && echo "create-jdbc-connection-pool --datasourceclassname org.postgresql.ds.PGConnectionPoolDataSource  --restype javax.sql.ConnectionPoolDataSource --property user=postgres:password=:DatabaseName=fin_dev:ServerName=$HOST_IP:port=5432 fing-pool" > mycommands.asadmin &&   cat ${DEPLOY_COMMANDS} >> mycommands.asadmin &&   ${PAYARA_PATH}/bin/asadmin start-domain -v --postbootcommandfile mycommands.asadmin ${PAYARA_DOMAIN}'
```

**Comment by lsantos** [comment]
Sweet guys, thanks for the prompt reply and the code samples.

I will definitely try out these new ideas and choose which one is more suitable for my requirements at hand.

**Comment by webelcomau** [comment]
Thanks @OndrejM and Payara Team for this awesome [postbootcommandfile](url) feature and work on Payara for Docker !

There is a minor type in your example from 26 May 2017, you have 'ENTRYPOINT=' (with an unnecessary =).

The following worked for me just using a mycommands.asadmin file to set the context and a context param after deployment:

```
FROM payara/server-full
...
COPY Test-1.0-SNAPSHOT.war $DEPLOY_DIR
COPY mycommands.asadmin .
USER root
RUN chown payara:payara mycommands.asadmin
USER payara

ENTRYPOINT ${PAYARA_PATH}/generate_deploy_commands.sh && cat ${DEPLOY_COMMANDS} >> mycommands.asadmin && ${PAYARA_PATH}/bin/asadmin start-domain -v --postbootcommandfile mycommands.asadmin ${PAYARA_DOMAIN}
```

Then in mycommands.asadmin:

```
deploy --contextroot test /opt/payara41/deployments/Test-1.0-SNAPSHOT.war
set-web-context-param --name=javax.faces.PROJECT_STAGE --value=Production Test-1.0-SNAPSHOT

```

**Comment by mikecroft** [comment]
@webelcomau Typo? What typo? :wink: 

Thanks for the tip! Glad the dockerfile is useful for you, this kind of feedback is great! I'll close this now since I think the original issue was dealt with some time ago.

**Comment by ofspain** [comment]
You guys are the bomb!!!!... 
Just copy and pasted this into my dockerfile while replacing necessary parameter e.g hostname database name etc
ENTRYPOINT ${PAYARA_PATH}/generate_deploy_commands.sh && \
  echo 'create-jdbc-connection-pool --datasourceclassname org.postgresql.ds.PGConnectionPoolDataSource
  --restype javax.sql.ConnectionPoolDataSource --property user=postgres:password=:DatabaseName=fin_dev:ServerName=$HOST_IP:port=5432 fing-pool' > mycommands.asadmin && \
  cat ${DEPLOY_COMMANDS} >> mycommands.asadmin && \
  ${PAYARA_PATH}/bin/asadmin start-domain -v --postbootcommandfile mycommands.asadmin ${PAYARA_DOMAIN}

IT JUST WORKED LIKE CHARM....
I have been battling for days before coming across this article.....Thanks a million times buddies

**Comment by kaiamrhein** [comment]
I am struggling with this..all i get is "/bin/sh: /generate_deploy_commands.sh: not found" so I assume thet my ${PAYARA_PATH} is empty.. What is it supposed to contain, and where would such content be supposed to be set? Many thanks

**Comment by kaiamrhein** [comment]
Sorry...I was working with the micro version...i figured it out now
